### PR TITLE
442 backgammon

### PR DESCRIPTION
**Pull request by msomatt**

### CODE REVIEW

File: modules/game/src/main/BSONHandlers.scala [hunk 1]
```diff
@@ -117,25 +119,50 @@ object BSONHandlers {
...
+                  p1 = Pocket(p1.map(_.role).map(Role.BackgammonRole)),
+                  p2 = Pocket(p2.map(_.role).map(Role.BackgammonRole))
+                )
+              }
             )
           )
         case _ => sys.error(s"Pocket Data BSON reader not implemented for GameLogic: ${r.intD("l")}")
       }
 
+    //todo put this in SG Piece.scala
```
Comment (JamesHeppell) [inline modules/game/src/main/BSONHandlers.scala:146]: will refactor this

File: bin/mongodb/kpis.js
Comment (msomatt) [inline bin/mongodb/kpis.js:1]: need to add nackgammon here?

File: modules/game/src/main/BSONHandlers.scala [hunk 2]
```diff
@@ -117,8 +119,25 @@ object BSONHandlers {
                   p1 = Pocket(p1.map(_.role).map(Role.GoRole)),
                   p2 = Pocket(p2.map(_.role).map(Role.GoRole))
                 )
-              },
-              promoted = r.str("t").view.flatMap(go.Pos.piotr).to(Set)
+              }
+            )
+          )
+        //TODO consider what we want to read out for backgammon pocket
```
Comment (msomatt) [inline modules/game/src/main/BSONHandlers.scala:125]: Is there anything to be done here?

File: modules/game/src/main/BSONHandlers.scala [hunk 3]
```diff
@@ -555,6 +576,58 @@ object BSONHandlers {
...
+            situation = backgammon.Situation(
+              backgammon.Board(
+                pieces = BinaryFormat.piece
+                  .readBackgammon(r bytes F.binaryPieces, gameVariant)
+                  .filterNot { case (_, posInfo) => posInfo._2 == 0 },
+                history = backgammon.History(
+                  lastTurn = turnUcis(r strO F.historyLastTurn),
+                  currentTurn = turnUcis(r strO F.historyCurrentTurn),
+                  //we can flatten as backgammon does not have any multiaction games
+                  //TODO: Is halfMoveClock even doing anything for backgammon?
```
Comment (msomatt) [inline modules/game/src/main/BSONHandlers.scala:599]: i expect this calculation is wrong, but i think its wrong for many game logics. the comment about flatten is untrue as well

File: modules/game/src/main/Event.scala [hunk 4]
```diff
@@ -306,18 +350,209 @@ object Event {
...
+            situation.dropsByRole
+          case _ => None
+        },
+        possibleLifts = situation match {
+          case (Situation.Backgammon(_)) => Some(situation.lifts.map(_.pos))
+          case _                         => None
+        },
+        pocketData = pocketData,
+        //TODO future multiaction games may not end turn on the same action, and this will need to be fixed
+        couldNextActionEndTurn = situation.actions.headOption.map(_ match {
```
Comment (msomatt) [inline modules/game/src/main/Event.scala:460]: is this correct for Backgammon?

File: modules/game/src/main/Game.scala [hunk 5]
```diff
@@ -323,7 +328,10 @@ case class Game(
     history.lastAction map {
       case d: Uci.Drop          => s"${d.pos}${d.pos}"
       case m: Uci.Move          => m.keys
+      case l: Uci.Lift          => s"^${l.pos}"
```
Comment (msomatt) [inline modules/game/src/main/Game.scala:331]: i think this wants to be the same as drop? its not uci its keys which i think are used for working out which squares to highlight

File: modules/round/src/main/JsonView.scala [hunk 6]
```diff
@@ -434,6 +447,9 @@ final class JsonView(
       }
     }
 
+  private def possibleLifts(pov: Pov): Option[JsValue] =
+    (pov.game playableBy pov.player) option { JsString(pov.game.situation.lifts.map(_.pos.key).mkString) }
+
   private def multiActionMetaData(pov: Pov): Option[JsObject] = {
     pov.game.variant.key match {
       case "monster" => multiActionMetaJson(pov)
```
Comment (msomatt) [inline modules/round/src/main/JsonView.scala:455]: do we need to add backgammon here?

File: modules/round/src/main/JsonView.scala [hunk 7]
```diff
@@ -445,9 +461,11 @@ final class JsonView(
   private def multiActionMetaJson(pov: Pov): Option[JsObject] = {
     //TODO future multiaction games may not end turn on the same action, and this will need to be fixed
     pov.game.situation.actions.headOption.flatMap(_ match {
-      case m: StratMove => Some(Json.obj("couldNextActionEndTurn" -> m.autoEndTurn))
-      case d: StratDrop => Some(Json.obj("couldNextActionEndTurn" -> d.autoEndTurn))
-      case _            => None
+      case m: StratMove    => Some(Json.obj("couldNextActionEndTurn" -> m.autoEndTurn))
+      case d: StratDrop    => Some(Json.obj("couldNextActionEndTurn" -> d.autoEndTurn))
+      case l: StratLift    => Some(Json.obj("couldNextActionEndTurn" -> l.autoEndTurn))
```
Comment (msomatt) [inline modules/round/src/main/JsonView.scala:466]: do we need to add DiceRoll here?

File: translation/source/variantShortName.xml [hunk 8]
```diff
@@ -34,8 +34,10 @@
   <string name="flipello10">XFlip</string>
   <string name="amazons">Amazons</string>
   <string name="oware">Oware</string>
-  <string name="togyzkumalak">Togyz</string>
+  <string name="togyzkumalak">Toqyz</string>
   <string name="go9x9">Go9</string>
   <string name="go13x13">Go13</string>
   <string name="go19x19">Go19</string>
+  <string name="backgammon">Backgammon</string>
```
Comment (msomatt) [inline translation/source/variantShortName.xml:41]: I think these are slightly too long by comparison to the others. Suggest Backg and Nackg?

File: ui/stratutils/src/main.ts [hunk 9]
```diff
@@ -123,7 +133,77 @@ const noFishnetVariants: VariantKey[] = [
...
+    case 'flipello10':
+      return 'flipello10';
+    case 'amazons':
+      return 'amazons';
+    case 'oware':
+      return 'oware';
+    case 'togyzkumalak':
+      return 'togyzkumalak';
+    default:
+      return 'chess';
```
Comment (msomatt) [inline ui/stratutils/src/main.ts:207]: i dont see backgammon in here?

**Pull request merged**